src/server/journal_test.cc

```diff
@@ -0,0 +1,124 @@
+#include <string>
+
+#include "base/gtest.h"
+#include "base/logging.h"
+#include "server/journal/serializer.h"
+#include "server/journal/types.h"
+
+using namespace testing;
+using namespace std;
+using namespace util;
+
+namespace dfly {
+
+struct EntryPayloadVisitor {
+  void operator()(const string_view sv) {
+    *out += sv;
+    *out += ' ';
+  }
+
+  void operator()(const ArgSlice slice) {
+    for (auto arg : slice) {
+      *out += arg;
+      *out += ' ';
+    }
+  }
+
+  void operator()(const pair<string_view, ArgSlice> p) {
+    (*this)(p.first);
+    (*this)(p.second);
+  }
+
+  void operator()(const CmdArgList list) {
+    for (auto arg : list) {
+      *out += facade::ToSV(arg);
+      *out += ' ';
+    }
+  }
+
+  void operator()(monostate) {
+  }
+
+  string* out;
+};
+
+// Extract payload from entry in string form.
+std::string ExtractPayload(journal::EntryNew& entry) {
+  std::string out;
+  EntryPayloadVisitor visitor{&out};
+  if (entry.owned_payload) {
+    CmdArgList list{entry.owned_payload->data(), entry.owned_payload->size()};
+    visitor(list);
+  } else {
+    std::visit(visitor, entry.payload);
+  }
+
+  if (out.size() > 0 && out.back() == ' ')
+    out.pop_back();
+
+  return out;
+}
+
+// Mock non-owned types with underlying storage.
+using StoredSlices = vector<vector<string_view>>;
+using StoredLists = vector<pair<vector<string>, CmdArgVec>>;
+
+template <typename... Ss> ArgSlice StoreSlice(StoredSlices* vec, Ss... strings) {
+  vec->emplace_back(initializer_list<string_view>{strings...});
+  return ArgSlice{vec->back().data(), vec->back().size()};
+}
+
+template <typename... Ss> CmdArgList StoreList(StoredLists* vec, Ss... strings) {
+  vector<string> stored_strings{strings...};
+  CmdArgVec out;
+  for (auto& s : stored_strings) {
+    out.emplace_back(s.data(), s.size());
+  }
+
+  vec->emplace_back(std::move(stored_strings), std::move(out));
+  auto& arg_vec = vec->back().second;
+  return CmdArgList{arg_vec.data(), arg_vec.size()};
+}
+
+// Test serializing and de-serializing entries.
+TEST(Journal, WriteRead) {
```

Comment:
The stat might look like much but the tests need some utils to mock dragonfly slice types